kubernetes/cache/cache.go

```diff
@@ -30,369 +16,148 @@ import (
-
-	c.stopClusterScopedChan = make(chan struct{})
-	kubeInformerFactory.Start(c.stopClusterScopedChan)
-	istioInformerFactory.Start(c.stopClusterScopedChan)
-	gatewayInformerFactory.Start(c.stopClusterScopedChan)
-
-	log.Info("[Kiali Cache] Waiting for cluster-scoped cache to sync")
-	for _, synced := range kubeInformerFactory.WaitForCacheSync(c.stopClusterScopedChan) {
-		if !synced {
-			log.Error("[Kiali Cache] failed to sync kube cluster-scoped cache")
-			return false
-		}
-	}
-	for _, synced := range istioInformerFactory.WaitForCacheSync(c.stopClusterScopedChan) {
-		if !synced {
-			log.Error("[Kiali Cache] failed to sync Istio cluster-scoped cache")
-			return false
-		}
-	}
-	for _, synced := range gatewayInformerFactory.WaitForCacheSync(c.stopClusterScopedChan) {
-		if !synced {
-			log.Error("[Kiali Cache] failed to sync Istio cluster-scoped cache")
-			return false
-		}
-	}
-
-	log.Info("[Kiali Cache] started")
-
-	return true
-}
-
-// CheckNamespace will
-// - Validate if a namespace is included in the cache
-// - Create and initialize a cache
-func (c *kialiCacheImpl) CheckNamespace(namespace string) bool {
-	if c.clusterScoped {
-		return true
-	} else if !c.isCached(namespace) {
-		return false
-	}
-
-	c.cacheLock.RLock()
-	_, isNsCached := c.nsCacheLister[namespace]
-	c.cacheLock.RUnlock()
-
-	if !isNsCached {
-		defer c.cacheLock.Unlock()
-		c.cacheLock.Lock()
-		return c.createNSCache(namespace)
-	}
-
-	return true
+// GetKubeCaches returns a kube cache for every configured Kiali Service Account client keyed by cluster name.
+func (c *kialiCacheImpl) GetKubeCaches() map[string]KubeCache {
+	return c.kubeCache
 }
 
-// Refresh will recreate the necessary cache. If the cache is cluster-scoped the "namespace" argument
-// is ignored and the whole cache is recreated, otherwise only the namespace-specific cache is updated.
-func (c *kialiCacheImpl) Refresh(namespace string) {
-	defer c.cacheLock.Unlock()
-	c.cacheLock.Lock()
+// Stops all caches across all clusters.
+func (c *kialiCacheImpl) Stop() {
+	log.Infof("Stopping Kiali Cache")
 
-	if c.clusterScoped {
-		c.stop("")
-		c.createClusterScopedCache()
-		return
+	wg := sync.WaitGroup{}
+	for _, kc := range c.kubeCache {
+		wg.Add(1)
+		go func(c KubeCache) {
+			defer wg.Done()
+			c.Stop()
+		}(kc)
 	}
+	wg.Wait()
 
-	c.stop(namespace)
-	c.createNSCache(namespace)
+	c.cleanup()
 }
 
-// refreshCache watches for changes to the cache's service account token
-// and recreates the cache(s) when the token changes.
-func (c *kialiCacheImpl) refreshCache(ctx context.Context, istioConfig rest.Config) {
-	ticker := time.NewTicker(60 * time.Second)
+// watchForClientChanges watches for changes to the cache's service account client
+// and recreates the cache(s) when the client changes. The client is updated when
+// the token for the client changes.
+func (c *kialiCacheImpl) watchForClientChanges(ctx context.Context, token string) {
```

Comment:
So, only the "Home cluster" client is updated. Perhaps it isn't clear in this PR, but I guess this is because the bound SA token only applies to the "local" cluster, while for remote ones we will be using standard non-expiring tokens.

Nothing to fix, just pointing out this.

Reply:
Yeah for remote clusters, I don't think the token will expire and even if it did the auto-renew mechanism that exists for the in-cluster token wouldn't auto-renew the out of cluster tokens. That being said, it could be good to watch for client changes for the other clients anyways in case the secrets change. I think that's something we can add later though when more of the pieces have come together and we can test this in earnest and have a clearer picture of the requirements around that. For now, just watching the client changes for the home cluster keeps the existing functionality of token renewal in single cluster which is why I added this.